### ISSUE TITLE
Rate limiting by default

### ISSUE THREAD

**Issue by realprogrammersusevim** [scsibug/nostr-rs-relay#52]
Over the past few days people have been testing out the security of Nostr and
seeing how easy it would be to take down large portions of the network. One of
the biggest problems is that anyone can create a massive number of pubkeys and
flood a relay with messages. Could this relay implementation add rate limiting
by default or at least notify users that they are completely open to spamming
and getting taken down? I believe Nostream has already done this. Or is there a
better way to deal with this issue to avoid being DDoSd?

**Comment by scsibug** [comment]
We have rate limiting, although it doesn't default to being enabled.  That is a great way to ensure that you can control the rate of growth of events on your relay.

If you want to avoid spam entirely, you have to do more than just rate limit, you have to authenticate/whitelist users in some way.  NIP-05 is the mechanism we currently have in this relay; and I think that will go a long way to controlling random spam.

I will be implementing this feature (https://todo.sr.ht/~gheartsfield/nostr-rs-relay/46) for extensibility, allowing the operator to have more control, and other applications to be able to take on the spam problem.

Ultimately, this is a policy decision, and there is a lot of latitude that each relay operator may want to have - making their relay private, NIP-05 only, or open to the world (with reasonable rate limits).  Network policy is also an option, although out of scope of the relay, IPs can be banned or rate limited, connections throttled, etc.

So, in summary; we have rate limiting, but that's up for an operator to decide on a value for.  We have NIP-05 validation in both passive and enforced mode, and NIP-05 whitelist/blacklists.  That's not going to be enough, but hopefully gRPC extensibility will allow other creative/dynamic solutions to take over.

I haven't looked at the specifics of what nostream is doing, but I'm sure there are lots of good ideas there.  Rate limiting on individual connections or pubkeys is a losing battle though; which is why I decided to make it global.

**Comment by realprogrammersusevim** [comment]
Thanks for the quick response, the change in https://todo.sr.ht/~gheartsfield/nostr-rs-relay/46 looks great. I had seen that you had implemented rate limiting when I went through the options. What I meant was displaying more prominently to the relay operator that their relay was not rate limited and the consequences of that and/or putting a rate limit as a default setting in the config.toml. The pentester was talking last night on a Nostr space that a lot of the major relays using nostr-rs-relay did not have any rate limiting at all.

**Comment by scsibug** [comment]
Good point.  I'm not against making rate limiting enabled by default, in fact that's probably wise.

I'm going to add a default limit per your suggestion, using 5/sec as a starting point: 18138106cba49f92498a80a0e6a7fe0d09cdd014
